README.md
# redux-api-unrest

[![Build Status](https://travis-ci.org/Kozea/redux-api-unrest.svg?branch=master)](https://travis-ci.org/Kozea/redux-api-unrest)
[![Coverage Status](https://coveralls.io/repos/github/Kozea/redux-api-unrest/badge.svg?branch=master)](https://coveralls.io/github/Kozea/redux-api-unrest?branch=master)

A redux reducer for unrest.


CHANGELOG.md
### [0.9.4](https://github.com/Kozea/redux-api-unrest/compare/v0.9.3...v0.9.4)

- Remove unused es-lint-react-plugin
- Revert @babel/runtime to dependencies

### [0.9.3](https://github.com/Kozea/redux-api-unrest/compare/v0.9.2...v0.9.3)

- Update deps
- Fix babel-runtime error when version differs from parent project

### [0.9.2](https://github.com/Kozea/redux-api-unrest/compare/v0.9.1...v0.9.2)

- Handle content-type parameters and missing content-type

### [0.9.1](https://github.com/Kozea/redux-api-unrest/compare/v0.9.0...v0.9.1)

- Update deps

## [0.9.0](https://github.com/Kozea/redux-api-unrest/compare/v0.8.4...v0.9.0)

- `rootPath` is now a function for runtime configurations
- Keep promises on instance for server rendering wait

### [0.8.4](https://github.com/Kozea/redux-api-unrest/compare/v0.8.3...v0.8.4)

- With sane query stringifying options

### [0.8.3](https://github.com/Kozea/redux-api-unrest/compare/v0.8.2...v0.8.3)

- With a working runtime

### [0.8.2](https://github.com/Kozea/redux-api-unrest/compare/v0.8.1...v0.8.2)

- Add non-json response handling for `fetchHandler`
- Replace `query-string` by `qs`
- Use `@babel/plugin-transform-runtime`

### [0.8.1](https://github.com/Kozea/redux-api-unrest/compare/v0.8.0...v0.8.1)

- Update deps

## [0.8.0](https://github.com/Kozea/redux-api-unrest/compare/v0.7.4...v0.8.0)

- Use fetch AbortController when available. Polyfill otherwise.

### [0.7.4](https://github.com/Kozea/redux-api-unrest/compare/v0.7.3...v0.7.4)

- Do not reset after fetch abort to prevent unecessary rendering

### [0.7.3](https://github.com/Kozea/redux-api-unrest/compare/v0.7.2...v0.7.3)

- Prevent crash with safari private mode for localStorage

### [0.7.2](https://github.com/Kozea/redux-api-unrest/compare/v0.7.1...v0.7.2)

- Update deps.

### [0.7.1](https://github.com/Kozea/redux-api-unrest/compare/v0.7.0...v0.7.1)

- Add fetch `payload` and restrict `parameters` to url parameters in `metadata`.

## [0.7.0](https://github.com/Kozea/redux-api-unrest/compare/v0.6.4...v0.7.0)

- Add fetch `url`, `method` and `parameters` in `metadata` in store when loading (FETCH action).
- Update deps.

### [0.6.4](https://github.com/Kozea/redux-api-unrest/compare/v0.6.3...v0.6.4)

- Update deps.

### [0.6.3](https://github.com/Kozea/redux-api-unrest/compare/v0.6.2...v0.6.3)

- Fix force bug when request got cached after abort.

### [0.6.2](https://github.com/Kozea/redux-api-unrest/compare/v0.6.1...v0.6.2)

- Don't try to merge objects if there is no objects in response.

### [0.6.1](https://github.com/Kozea/redux-api-unrest/compare/v0.6.0...v0.6.1)

- Dependencies update

## [0.6.0](https://github.com/Kozea/redux-api-unrest/compare/v0.5.3...v0.6.0)

- Add abort on reset.
- Add .force methods to force abort on previously loading endpoints.

### [0.5.3](https://github.com/Kozea/redux-api-unrest/compare/v0.5.2...v0.5.3)

- Add a reset action.

### [0.5.2](https://github.com/Kozea/redux-api-unrest/compare/v0.5.1...v0.5.2)

- Add json content on http error when response is json.

### [0.5.1](https://github.com/Kozea/redux-api-unrest/compare/v0.5.0...v0.5.1)

- Fix various packaging problems.

## [0.5.0](https://github.com/Kozea/redux-api-unrest/compare/v0.4.0...v0.5.0)

- Use uppercase methods.

## [0.4.0](https://github.com/Kozea/redux-api-unrest/compare/v0.3.2...v0.4.0)

- Replace `methodAll()` functions with `method()` and `method()` with `methodItem()`.

### [0.3.2](https://github.com/Kozea/redux-api-unrest/compare/v0.3.1...v0.3.2)

- Add fetch result / error as promise resolve.

### [0.3.1](https://github.com/Kozea/redux-api-unrest/compare/v0.3.0...v0.3.1)

- Patch for bad utils import.

## [0.3.0](https://github.com/Kozea/redux-api-unrest/compare/v0.2.3...v0.3.0)

- Add support for query parameters as payload argument for get method.


test/setup.js
// AbortController is just a mockup. Testing with the polyfill
delete global.AbortController


test/utils.js
export const timeout = time => new Promise(resolve => setTimeout(resolve, time))


test/misc.test.js
import { applyMiddleware, combineReducers, createStore } from 'redux'
import thunk from 'redux-thunk'
// eslint-disable-next-line no-unused-vars
import regeneratorRuntime from 'regenerator-runtime'

import ApiUnrest from '../src'
import { timeout } from './utils'

describe('Api unrest can handle JWT', () => {
  it('sends the token on fetch request', async () => {
    const storage = {
      getItem: name => storage[name],
      setItem: (name, value) => (storage[name] = value),
      removeItem: name => delete storage[name],
      jwt: 'JWTTOKEN',
    }

    const api = new ApiUnrest(
      {
        fruit: 'fruit',
        color: 'base/color/:id?',
        tree: 'forest/tree/:type?/:age?',
      },
      {
        JWTStorage: storage,
        fetch: async (url, opts) => {
          await timeout(25)
          return {
            status: 200,
            headers: {
              get: key =>
                ({
                  'Content-Type': 'application/json',
                }[key]),
            },
            // eslint-disable-next-line require-await
            json: async () => ({
              objects: [{ headers: opts.headers }],
            }),
          }
        },
      }
    )
    const store = createStore(
      combineReducers(api.reducers),
      applyMiddleware(thunk)
    )
    await store.dispatch(api.actions.color.get())
    expect(store.getState().color.objects[0].headers.Authorization).toEqual(
      'Bearer JWTTOKEN'
    )
  })
  it('sets the token on fetch response', async () => {
    const storage = {
      getItem: name => storage[name],
      setItem: (name, value) => (storage[name] = value),
      removeItem: name => delete storage[name],
      jwt: 'OLDTOKEN',
    }
    const api = new ApiUnrest(
      {
        fruit: 'fruit',
        color: 'base/color/:id?',
        tree: 'forest/tree/:type?/:age?',
      },
      {
        JWTStorage: storage,
        fetch: async () => {
          await timeout(25)
          return {
            status: 200,
            headers: {
              get: key =>
                ({
                  'Content-Type': 'application/json',
                  Authorization: 'JWTTOKEN',
                }[key]),
            },
            // eslint-disable-next-line require-await
            json: async () => ({}),
          }
        },
      }
    )
    await api.actions.color.get()(() => ({}))
    expect(storage.jwt).toEqual('JWTTOKEN')
  })
  it('removes the token on a 401', async () => {
    const storage = {
      getItem: name => storage[name],
      setItem: (name, value) => (storage[name] = value),
      removeItem: name => delete storage[name],
      jwt: 'shouldBeRemoved',
    }
    const api = new ApiUnrest(
      {
        fruit: 'fruit',
        color: 'base/color/:id?',
        tree: 'forest/tree/:type?/:age?',
      },
      {
        JWTStorage: storage,
        fetch: async () => {
          await timeout(25)
          return {
            status: 401,
            headers: {
              get: key =>
                ({
                  'Content-Type': 'application/json',
                  Authorization: 'BADJWTTOKEN',
                }[key]),
            },
            // eslint-disable-next-line require-await
            json: async () => ({}),
          }
        },
      }
    )
    try {
      await api.actions.color.get()(() => ({}))
    } catch (err) {
      expect()
    }
    expect(storage.jwt).toBeUndefined()
  })
  it('uses the localStorage if available', () => {
    global.localStorage = { iam: 'localStorage' }
    const api = new ApiUnrest({}, { JWTStorage: true })
    delete global.localStorage
    expect(api.storage).toEqual({ iam: 'localStorage' })
  })
  it('do nothing if the localStorage is unavailable', () => {
    const api = new ApiUnrest({}, { JWTStorage: true })
    expect(api.storage).toBeNull()
  })
})


src/adapter.js
// This import may or may not be a good idea.
// It's needed but can be included by your app.
// eslint-disable-next-line no-unused-vars
import regeneratorRuntime from 'regenerator-runtime'

const storage = typeof localStorage == 'undefined' ? null : localStorage

export const removeTrailingSlash = url => {
  const [path, search] = url.split('?', 2)
  if (search) {
    return [path.replace(/\/+$/, ''), search].join('?')
  }
  return path.replace(/\/+$/, '')
}

export default async (url, opts) => {
  opts.method = opts.method || 'GET'
  opts.headers = {
    Accept: 'application/json',
  }
  const jwt = storage && storage.getItem('jwt')
  if (jwt) {
    opts.headers.Authorization = `Bearer ${jwt}`
  }
  if (opts.method !== 'GET') {
    opts.headers['Content-Type'] = 'application/json'
    opts.body = JSON.stringify(opts.body)
  }
  let response, json
  try {
    response = await fetch(removeTrailingSlash(url), opts)
  } catch (e) {
    throw new Error('Erreur lors de la récupération de données.')
  }
  if (response.status > 300 || response.status < 200) {
    // TODO: find a better solution. UNREST ?
    if (response.status === 404 && opts.method === 'GET') {
      return { occurences: 0, objects: [] }
    }
    if (response.status === 401 && storage) {
      storage.removeItem('jwt')
    }
    if (response.headers.get('Content-Type') !== 'application/json') {
      const text = await response.text()
      throw new Error(
        `Erreur ${response.status} lors de la récupération de données. ${text}`
      )
    }
    try {
      json = await response.json()
    } catch (e) {
      throw new Error('Erreur de lecture de l’erreur serveur. JSON invalide')
    }
    const newError = new Error(
      `Erreur ${response.status} serveur. ${json.message || json.description}`
    )
    json.method = opts.method || 'GET'
    newError.json = json
    throw newError
  }
  try {
    json = await response.json()
  } catch (e) {
    throw new Error('Erreur de lecture de la réponse serveur. JSON invalide')
  }
  if (response.headers.get('Authorization') && storage) {
    storage.setItem('jwt', response.headers.get('Authorization'))
  }
  return json
}


src/helpers.js
export const DEFAULT_METHODS = ['get', 'put', 'post', 'patch', 'delete']

// CRUD helper for redux-api. Allow the use of method like api.actions.put({})
// Usage:
// * dispatch(api.actions.endpoint.get(cb))
// * dispatch(api.actions.endpoint.get(params))
// * dispatch(api.actions.endpoint.post(obj, cb))
// * dispatch(api.actions.endpoint.patch(obj, params, cb))
export const crud = methods =>
  // For each of the provided methods
  (methods || DEFAULT_METHODS).reduce((map, name) => {
    // Returns the helper function for the method
    // TODO: Think about how improving the async there (promise + cb = :/)
    map[name] = (body, params, cb) => {
      // If the method does not take a body the argument list is (params, cb)
      if (name === 'get' || name === 'delete') {
        params = body
        cb = params
        body = void 0
      }
      // The params argument can be omitted
      if (params instanceof Function) {
        cb = params
        params = {}
      }

      return [params, { body, method: name.toUpperCase() }, cb || (() => {})]
    }
    return map
  }, {})


src/transformer.js
// Determines if two objects are equal according to the given pks
const pkEqual = pks => (o1, o2) => pks.every(pk => o1[pk] === o2[pk])

// A filter that only returns objects that are not in the given list
const notIn = pks => objects => object =>
  !objects.some(o => pkEqual(pks)(o, object))

// This is a reduxApi transformer that works with unrest responses
export default (data, olds, action) => {
  if (!data) {
    return []
  }
  const { objects, primary_keys: pks } = data

  // We add the number of occurences as an extra parameter
  action.request.extras = { occurences: data.occurences }
  const method = action.request.params ? action.request.params.method : ''
  switch (method) {
    case 'POST':
      // In case of a POST we concatenate the new data to the old
      return [...olds, ...objects]
    case 'PUT':
      // In case of a PUT we replace all if it's a batch
      if (!Object.keys(action.request.pathvars).length) {
        // If there's no path variables it's a batch PUT so
        return [...objects]
      }
    // In case of a PUT one we replace the one that changed
    // Since it's exactly like PATCH we are falling through
    // eslint-disable-next-line no-fallthrough
    case 'PATCH':
      // In case of a PATCH we replace the element that changed
      // Old objects without the updated new objects + the new objects
      return [...olds.filter(notIn(pks)(objects)), ...objects]
    case 'DELETE':
      // In case of a DELETE we remove all
      return [...olds.filter(notIn(pks)(objects))]
  }

  return objects
}


src/index.js
import deepEqual from 'deep-equal'
import isoFetch from 'isomorphic-fetch'
import { compile } from 'path-to-regexp'
import { stringify } from 'qs'

import { AbortController, patchFetchMaybe } from './ponyfill'
import { isEmpty } from './utils'

export const apiUnrestPrefix = '@@api-unrest'
export const methods = ['get', 'put', 'post', 'patch', 'delete']
export const initialEndpointState = {
  objects: [],
  metadata: {},
  loading: false,
  error: null,
  lastFetch: null,
  lastFetchParameters: null,
}

export const httpError = (code, description, json) => {
  const error = new Error(`[${code}] - ${description}`)
  error.name = 'HttpError'
  error.code = code
  error.description = description
  error.json = json
  return error
}

export default class ApiUnrest {
  constructor(
    routes,
    // The following prefix must be the redux mount point to use cache
    options
  ) {
    // Set default options
    options = {
      prefix: 'api',
      rootPath: () => '',
      cache: null,
      JWTStorage: false,
      errorHandler: () => true,
      apiRoot: state => state[this.prefix],
      fetch: isoFetch,
      ...options,
    }
    this.prefix = options.prefix
    this.rootPath = options.rootPath
    this.cache = options.cache
    this.storage = null // eslint-disable-next-line no-unused-vars
    this.apiRoot = options.apiRoot
    this.errorHandler = options.errorHandler
    if (options.JWTStorage) {
      if (options.JWTStorage === true) {
        if (typeof localStorage !== 'undefined') {
          this.storage = localStorage
        }
      } else {
        this.storage = options.JWTStorage
      }
    }
    this.fetch = patchFetchMaybe(options.fetch)

    this.routes = { ...routes }
    this.fetches = {}
    this.promises = {}
    this.events = this.getEvents(routes)
    this.actions = this.getActions(routes)
    this.reducers = this.getReducers(routes)
  }

  getEvents(routes) {
    return Object.keys(routes).reduce((events, endpoint) => {
      const eventPath = `${apiUnrestPrefix}/${this.prefix}/${endpoint}`
      events[endpoint] = {
        fetch: `${eventPath}/FETCH`,
        success: `${eventPath}/SUCCESS`,
        error: `${eventPath}/ERROR`,
        abort: `${eventPath}/ABORT`,
        cache: `${eventPath}/CACHE`,
        reset: `${eventPath}/RESET`,
      }
      return events
    }, {})
  }

  getActions(routes) {
    // For each routes return a map: endpoint -> methods -> thunk -> fetch
    return Object.entries(routes).reduce((actions, [endpoint, path]) => {
      const urlFormatter = compile(`/${path}`)
      actions[endpoint] = methods.reduce(
        (routeActions, method) => {
          routeActions[method] = (payload = {}, force = false) =>
            this.fetchThunk(
              endpoint,
              urlFormatter,
              {},
              method.toUpperCase(),
              payload,
              force
            )
          routeActions[`${method}Item`] = (
            urlParameters,
            payload = {},
            force = false
          ) => {
            if (!urlParameters || isEmpty(urlParameters)) {
              throw new Error(
                `${method}Item on ${
                  this.prefix
                }.${endpoint} called without parameters`
              )
            }
            return this.fetchThunk(
              endpoint,
              urlFormatter,
              urlParameters,
              method.toUpperCase(),
              payload,
              force
            )
          }
          routeActions.force[method] = (payload = {}) =>
            routeActions[method](payload, true)
          routeActions.force[`${method}Item`] = (urlParameters, payload = {}) =>
            routeActions[`${method}Item`](urlParameters, payload, true)
          return routeActions
        },
        { force: {} }
      )
      actions[endpoint].reset = () => dispatch => {
        if (this.fetches[endpoint]) {
          this.fetches[endpoint].abort()
        }
        dispatch({ type: this.events[endpoint].reset })
      }
      return actions
    }, {})
  }

  getReducers(routes) {
    return Object.keys(routes).reduce((reducers, endpoint) => {
      reducers[endpoint] = (state = initialEndpointState, action) => {
        switch (action.type) {
          case this.events[endpoint].fetch:
            return {
              ...state,
              loading: true,
              error: null,
              metadata: action.metadata,
            }
          case this.events[endpoint].success:
            return {
              ...state,
              objects: action.objects
                ? this.mergeObjects(
                    action.method,
                    state.objects,
                    action.objects,
                    action.metadata.primary_keys,
                    action.batch
                  )
                : state.objects,
              metadata: action.metadata,
              loading: false,
              error: null,
              lastFetch: action.method === 'GET' ? Date.now() : state.lastFetch,
              lastFetchParameters:
                action.method === 'GET'
                  ? action.parameters
                  : state.lastFetchParameters,
            }
          case this.events[endpoint].error:
            return {
              ...state,
              loading: false,
              error: action.error,
            }
          case this.events[endpoint].abort:
            return {
              ...state,
              loading: false,
              error: null,
            }
          case this.events[endpoint].cache:
            return {
              ...state,
              loading: false,
            }
          case this.events[endpoint].reset:
            return {
              ...initialEndpointState,
            }
          default:
            return state
        }
      }
      return reducers
    }, {})
  }

  mergeObjects(method, olds, objects, pks, batch) {
    // An equality based on primary keys
    const pkEqual = (o1, o2) => pks.every(pk => o1[pk] === o2[pk])
    // A filter that only returns objects that are not in the given list
    const notIn = objs => obj => !objs.some(o => pkEqual(o, obj))
    switch (method) {
      case 'GET':
        // In case of a GET we simply use the new objects
        return [...objects]
      case 'POST':
        // In case of a POST we concatenate the new data to the old
        return [...olds, ...objects]
      case 'PUT':
        // In case of a PUT we replace all if it's a batch
        if (batch) {
          // If there's no path variables it's a batch PUT so
          return [...objects]
        }
      // In case of a PUT one we replace the one that changed
      // Since it's exactly like PATCH we are falling through
      // eslint-disable-next-line no-fallthrough
      case 'PATCH':
        // In case of a PATCH we replace the element that changed
        // Old objects without the updated new objects + the new objects
        return [...olds.filter(notIn(objects)), ...objects]
      case 'DELETE':
        // In case of a DELETE we remove all
        return [...olds.filter(notIn(objects))]
    }
  }

  fetchThunk(endpoint, urlFormatter, urlParameters, method, payload, force) {
    return async (dispatch, getState) => {
      const query =
        method === 'GET' && !isEmpty(payload)
          ? stringify(payload, {
              addQueryPrefix: true,
              strictNullHandling: true,
              indices: false,
            })
          : ''
      const url = this.rootPath() + urlFormatter(urlParameters) + query
      // In case of a get request, add get parameters to parameters
      // (This prevents cache on different url queries: ?offset=0 vs ?offset=10)
      const parameters =
        method === 'GET' ? { ...urlParameters, ...payload } : urlParameters
      const handleError = error => {
        // If error handler returns true, the error will propagate
        if (
          this.errorHandler(
            error,
            {
              endpoint,
              url,
              urlParameters,
              method,
              payload,
              prefix: this.prefix,
            },
            dispatch,
            getState
          )
        ) {
          throw error
        }
      }
      const state = getState()
      const { loading } = this.apiRoot(state)[endpoint]
      if (loading) {
        if (force) {
          this.fetches[endpoint].abort()
        } else {
          // Waiting for AbortController api
          // https://developer.mozilla.org/en-US/docs/Web/API/AbortController
          const alreadyLoadingError = new Error('Already loading')
          alreadyLoadingError.name = 'AlreadyLoadingError'
          handleError(alreadyLoadingError)
          return { status: 'failed', error: alreadyLoadingError }
        }
      }
      // Here we go
      dispatch({
        type: this.events[endpoint].fetch,
        metadata: {
          url,
          method,
          parameters: urlParameters,
          payload,
        },
      })
      // Prevent cache on forced request
      if (!loading && !force && this.cache && method === 'GET') {
        const endpointState = this.apiRoot(state)[endpoint]
        const { lastFetch, lastFetchParameters } = endpointState
        if (
          lastFetch &&
          Date.now() - lastFetch < this.cache &&
          (lastFetchParameters === null ||
            deepEqual(lastFetchParameters, parameters))
        ) {
          dispatch({ type: this.events[endpoint].cache })
          return {
            status: 'cache',
            objects: endpointState.objects,
            metadata: endpointState.metadata,
          }
        }
      }
      try {
        this.fetches[endpoint] = new AbortController()
        this.promises[endpoint] = this.fetchHandler(
          url,
          method,
          payload,
          this.fetches[endpoint].signal
        )
        const { objects, ...metadata } = await this.promises[endpoint]
        dispatch({
          type: this.events[endpoint].success,
          objects,
          metadata,
          method,
          parameters,
          // This request was a batch if there were no url parameters
          batch: isEmpty(urlParameters),
        })
        return { status: 'success', objects, metadata }
      } catch (error) {
        if (error.name === 'AbortError') {
          dispatch({
            type: this.events[endpoint].abort,
          })
          return { status: 'aborted' }
        }
        dispatch({
          type: this.events[endpoint].error,
          error: error.toString(),
        })
        handleError(error)
        return { status: 'failed', error }
      } finally {
        delete this.fetches[endpoint]
        delete this.promises[endpoint]
      }
    }
  }

  async fetchHandler(url, method, payload, signal) {
    const opts = {
      method,
      headers: {
        Accept: 'application/json',
      },
      signal,
    }
    if (method !== 'GET' && !isEmpty(payload)) {
      opts.headers['Content-Type'] = 'application/json'
      opts.body = JSON.stringify(payload)
    }
    const response = await this._fetch(url, opts)
    const [contentType] = (response.headers.get('Content-Type') || '').split(
      ';'
    )

    if (response.status > 300 || response.status < 200) {
      if (contentType !== 'application/json') {
        const text = await response.text()
        throw httpError(response.status, text)
      }
      const json = await response.json()
      if (response.status === 404 && json.occurences === 0) {
        return json
      }
      throw httpError(response.status, json.message || json.description, json)
    }
    if (contentType !== 'application/json') {
      // This could happen if used outside of redux context
      return response.blob()
    }
    return { ...(await response.json()), code: response.status }
  }

  onBeforeFetchHook({ url, opts }) {
    if (this.storage) {
      let jwt = null
      try {
        jwt = this.storage.getItem('jwt')
      } catch (e) {
        // Safari private mode
        console && console.warn(e)
      }
      if (jwt) {
        opts.headers.Authorization = `Bearer ${jwt}`
      }
    }
    return { url, opts }
  }

  async _fetch(url, opts) {
    const hookParams = this.onBeforeFetchHook({ url, opts })
    const response = await this.fetch(hookParams.url, hookParams.opts)
    return this.onAfterFetchHook(hookParams, response)
  }

  // eslint-disable-next-line no-unused-vars
  onAfterFetchHook({ url, opts }, response) {
    if (this.storage) {
      if (response.status === 401) {
        try {
          this.storage.removeItem('jwt')
        } catch (e) {
          // Safari private mode
          console && console.warn(e)
        }
      } else if (response.headers.get('Authorization')) {
        try {
          this.storage.setItem('jwt', response.headers.get('Authorization'))
        } catch (e) {
          // Safari private mode
          console && console.warn(e)
        }
      }
    }
    return response
  }
}
